R/server-helpers.R

```diff
@@ -322,76 +329,73 @@ getPlotData <- function(prjdata, query, pltscen, diffscen, key, filtervar=NULL,
-        dp$region <- factor(dp$region, levels=c(names(region32), '0'), ordered=TRUE)
+        if('region' %in% names(dp)) {
+            dp$region <- factor(dp$region, levels=c(names(region32), '0'), ordered=TRUE)
+        }
     }
     else {
         dp <- NULL
     }
 
-    yearcols <- year.cols(tp)
     if(!is.null(dp)) {
-        ## we're doing a difference plot, so subtract the difference scenario.
+        ## We're doing a difference plot, so subtract the difference scenario.
         ## Join the data sets first so that we can be sure that we have matched
         ## the rows and columns correctly
         varnames <- names(tp)
-        mergenames <- varnames[!varnames %in% c('scenario', 'Units') &
-                                   !grepl(year.regex, varnames)]
+        mergenames <- varnames[!varnames %in% c('scenario', 'Units', 'value')]
+
         joint.data <- merge(tp, dp, by=mergenames, all=TRUE)
         if(anyNA(joint.data))
             joint.data[is.na(joint.data)] <- 0 # zero out missing values
 
-        ## find common years between the data sets
-        years.tp <- grep('^X[0-9]{4}\\.x',
-                         names(joint.data), value=TRUE) %>%
-            substr(1,5)        # get the year columns from tp and strip the '.x'
-        years.dp <- grep('^X[0-9]{4}\\.y',
-                         names(joint.data), value=TRUE) %>%
-            substr(1,5)                 # same for dp
-        yearcols <- intersect(years.tp, years.dp)
-
-        d1 <- as.matrix(joint.data[paste0(yearcols,'.x')])
-        d2 <- as.matrix(joint.data[paste0(yearcols,'.y')])
-        delta <- d1-d2
-        colnames(delta) <- yearcols
-
-        ## Construct the new data frame.  We use the scenario name from the left
-        ## (dp) data frame.
+        value <- joint.data$value.x - joint.data$value.y
+
+        mergenames <- sapply(mergenames, as.name) # Don't eval hyphenated col names
+
+        # Construct the new data frame.  We use the scenario name from the left
+        # (dp) data frame.
         tp <- dplyr::rename(joint.data, scenario=scenario.x, Units=Units.x) %>%
-            dplyr::select_(.dots=c('scenario', mergenames, 'Units')) %>% cbind(delta)
+           dplyr::select_(.dots=c('scenario', mergenames, 'Units')) %>% cbind(value)
     }
 
     ## If filtering is in effect, do it now
     if(!is.null(filtervar) &&
        !is.null(filterset) &&
-       length(filterset) > 0
+       length(filterset) > 0 &&
+       filtervar %in% names(tp)
        ) {
-        ## This is horrible.  There has got to be a better way to do this.
-        tp <- dplyr::filter_(tp, paste(filtervar,
-                                '%in% c(',
-                                paste(shQuote(filterset), collapse=', '), ')'
-                                ))
+
+        tp <- dplyr::filter_(tp, lazyeval::interp(~y %in% x, y = as.name(filtervar), x = filterset))
+
+        #if(nrow(tp) == 0) {
```

Comment:
Is this code still needed?  If not, then delete.  Don't leave commented-out code in the final product.